### DISCUSSION TITLE
Using with react query cause maxium depth re-render on http error.

### DISCUSSION THREAD

**Discussion by dante01yoon** [pmndrs/jotai#1520] [Q&A] [answered]
Hi, using with react query with `useAtom / useAtomValue` cause infinite re-render when http response has 4xx , 5xx.

Here is my example  codes. 

```tsx
const Component = () => {
  const {data} = useQuery(["someKey"], async () => { axios.get("something") 
    return {
      ...query,
      data: query.data ? query.data.data : []
    }
  });
 
const dataAtom =  useMemo( () => atom(data),[data]);
const dataValue = useAtomValue(dataAtom);
}
```

those codes works well  when http response is 2xx, but cause infinite re-render on 4xx, 5xx. 

I guess calling query repeatedly cause  re-creation of `dataAtom` and that cause datavalue re created ?  

If so, is calling `useAtom, useAtomValue` cause re-rendering?

**Comment by dante01yoon** [comment, marked as answer]
hmm, I don't know how react query implemented deeply, but It might caused by throw some promise deep inside in useQuery would cause this issues.

**Reply by dante01yoon** [reply]
one possible solution might be replace dependency arrays in useMemo with isError or isSuccess. This would not cause  update `dataValue`.
But this would be only valid solution with no more http response updated.
```tsx
const Component = () => {
  const {data, isError, isSuccess} = useQuery(["someKey"], async () => { axios.get("something") 
    return {
      ...query,
      data: query.data ? query.data.data : []
    }
  });
 
const dataAtom =  useMemo( () => atom(data),[isError, isSuccess]);
const dataValue = useAtomValue(dataAtom);
}
```

**Reply by dai-shi** [reply]
Thanks for the investigation.
Sounds like useQuery returns a new reference on every call. So, `[data]` deps don't work.
You would need something stable. What's `data` exactly when error occurs?

**Reply by dante01yoon** [reply]
```ts
 const {data, isError, isSuccess} = useQuery(["someKey"], async () => { axios.get("something") 
```
data would `undefined` when error occurs

**Reply by dai-shi** [reply]
Hmm, if it's `undefined`, `[data]` shouldn't cause infinite loop...

**Reply by dante01yoon** [reply]
Oh, sorry for the confusion. There was missing part of codes  for brevity.

In most of cases, we don't directly use `useQuery`  in component, however, do normalizing for components .
So  if we have  create custom response for http error, it create new reference for objects and would cause infinite rendering loop

```ts
const someQuery  = () => {
  return useQuery(["someKey"], async () => { axios.get("something") 
      return {
        ...query,
        data: query.data ? query.data.data : [] // <- return on http error response 
      }
    });
  }
```

```tsx
const Component = () => {
  const {data, isError, isSuccess} = someQuery()
 
const dataAtom =  useMemo( () => atom(data),[isError, isSuccess]);
const dataValue = useAtomValue(dataAtom);
}
```

in above codes, new array will be returned on every  refetching, so I wonder  it cause infinite render?

**Reply by dante01yoon** [reply]
here is simple examples  codesandbox  I created.

below code doesn't cause re-render
```
  const {data, isError, isSuccess} = someQuery()

```
this  code cause re-render
```
  const {data = [], isError, isSuccess} = someQuery()
```

https://codesandbox.io/s/new-frost-jh0tkz?file=/src/FetchComponent.js. 
----------
Oh the link can't even  open the page because infinite loop 😢

**Reply by dai-shi** [reply]
You can just define `[]` outside components/hooks and it should work.
```js
const EMPTY_DATA = []
```

**Reply by dante01yoon** [reply]
that would be one workaround. 

Or  just use get empty array from `useMemo` in useQuery. this would be better to guarantee  each hook use different empty array.

**Reply by dante01yoon** [reply]
Why this happened when use `useAtom` or `useAtomValue` ?  infinite loop doesn't occur when use `useState`. 🤔

**Reply by dai-shi** [reply]
Because they subscribe for the atom with `useEffect(..., [atom])`. Atoms are distinguished by references.

**Reply by dante01yoon** [reply]
Thanks for your answer. I'll have a more deep look into it

**Comment by dante01yoon** [comment]
for who already get struggled upon on this, 

It would be better write custom hook using useQuery, with useMemo.

```js
  const queryInfo = useQuery(["someKey"], async () => { axios.get("something") 
    return {
      ...query,
      data: query.data ? query.data.data : []
    }
  });
 
return useMemo(() => queryInfo.data, [queryInfo.data])

```
queryInfo changed every render,  but queryInfo.data is not. use should wrap data with useMemo to prevent additional re-render.
